File: Tree-shaped Deep Forest/Minimum Spanning Tree/Q1. Find Critical and Pseudo-Critical Edges in Minimum Spanning Tree/Q1_FindCriticalandPseudo-CriticalEdgesinMinimumSpanningTree.java
class UnionFind {
    private int[] parent;  // Array to store parent of each node

    public UnionFind(int n) {
        parent = new int[n];  // Initialize parent array of size n
        
        // Initially, each node is its own parent (separate components)
        for (int i = 0; i < n; i++)
            parent[i] = i;
    }

    public int findParent(int p) {
        // If p is its own parent, return p
        // Else recursively find root and apply path compression
        return parent[p] == p ? p : (parent[p] = findParent(parent[p]));
    }

    public void union(int u, int v) {
        // Find root parents of u and v
        int pu = findParent(u), pv = findParent(v);
        
        // Connect root of u to root of v (merge sets)
        parent[pu] = pv;
    }
}

class Solution {
    public List<List<Integer>> findCriticalAndPseudoCriticalEdges(int n, int[][] edges) {
        
        // List to store critical edges (must be in MST)
        List<Integer> critical = new ArrayList<>();
        
        // List to store pseudo-critical edges (can be in MST)
        List<Integer> pseudoCritical = new ArrayList<>();
        
        // Add index to each edge to track original position
        for (int i = 0; i < edges.length; i++) {
            int[] edge = edges[i];  // Get current edge
            
            // Increase size of edge array by 1 to store index
            edge = Arrays.copyOf(edge, edge.length + 1);
            
            // Store original index at position 3
            edge[3] = i;
            
            // Update edges array
            edges[i] = edge;
        }
        
        // Sort edges based on weight (index 2)
        Arrays.sort(edges, (a, b) -> Integer.compare(a[2], b[2]));

        // Find MST weight without blocking or forcing any edge
        int mstwt = findMST(n, edges, -1, -1);

        // Iterate through each edge
        for (int i = 0; i < edges.length; i++) {
            
            // If removing this edge increases MST weight → critical edge
            if (mstwt < findMST(n, edges, i, -1))
                critical.add(edges[i][3]);  // Add original index
            
            // If forcing this edge keeps MST weight same → pseudo-critical
            else if (mstwt == findMST(n, edges, -1, i))
                pseudoCritical.add(edges[i][3]);  // Add original index
        }

        // Prepare final result list
        List<List<Integer>> result = new ArrayList<>();
        
        result.add(critical);        // First list → critical edges
        result.add(pseudoCritical);  // Second list → pseudo-critical edges
        
        return result;  // Return result
    }

    private int findMST(int n, int[][] edges, int block, int e) {
        
        // Initialize Union-Find for n nodes
        UnionFind uf = new UnionFind(n);
        
        int weight = 0;  // Total weight of MST

        // If we want to force include edge 'e'
        if (e != -1) {
            weight += edges[e][2];  // Add its weight
            
            // Union its two vertices
            uf.union(edges[e][0], edges[e][1]);
        }

        // Iterate through all edges
        for (int i = 0; i < edges.length; i++) {
            
            // Skip the blocked edge
            if (i == block)
                continue;

            // If both nodes have same parent → cycle → skip
            if (uf.findParent(edges[i][0]) == uf.findParent(edges[i][1]))
                continue;

            // Otherwise include edge in MST
            uf.union(edges[i][0], edges[i][1]);
            
            // Add weight of this edge
            weight += edges[i][2];
        }

        // Check if all nodes are connected (valid MST)
        for (int i = 0; i < n; i++) {
            
            // If any node has different root → graph not fully connected
            if (uf.findParent(i) != uf.findParent(0))
                return Integer.MAX_VALUE;  // Return large value (invalid MST)
        }

        return weight;  // Return total MST weight
    }
}

File: Tree-shaped Deep Forest/Minimum Spanning Tree/Q1. Find Critical and Pseudo-Critical Edges in Minimum Spanning Tree/Q1_FindCriticalandPseudo-CriticalEdgesinMinimumSpanningTree.py
class UnionFind(object):
    def __init__(self, n):
        # Initially, each node is its own parent (separate components)
        self.parent = list(range(n))

    def find_parent(self, p):
        # If p is the root, return it
        if self.parent[p] == p:
            return p
        
        # Path compression: make the parent of p point directly to root
        self.parent[p] = self.find_parent(self.parent[p])
        return self.parent[p]

    def union(self, u, v):
        # Find root parents of u and v
        pu = self.find_parent(u)
        pv = self.find_parent(v)
        
        # Merge the two components
        self.parent[pu] = pv


class Solution(object):
    def findCriticalAndPseudoCriticalEdges(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: List[List[int]]
        """

        # Lists to store result
        critical = []
        pseudo_critical = []

        # Add index to each edge to track original position
        for i in range(len(edges)):
            edges[i].append(i)

        # Sort edges based on weight (3rd element)
        edges.sort(key=lambda x: x[2])

        # Find MST weight without any restriction
        mst_wt = self.find_mst(n, edges, -1, -1)

        # Check each edge
        for i in range(len(edges)):

            # Case 1: If removing this edge increases MST weight → Critical
            if mst_wt < self.find_mst(n, edges, i, -1):
                critical.append(edges[i][3])

            # Case 2: If forcing this edge keeps MST same → Pseudo-Critical
            elif mst_wt == self.find_mst(n, edges, -1, i):
                pseudo_critical.append(edges[i][3])

        return [critical, pseudo_critical]


    def find_mst(self, n, edges, block, e):
        # Create Union-Find structure
        uf = UnionFind(n)
        weight = 0  # Total MST weight

        # If we force include edge 'e'
        if e != -1:
            weight += edges[e][2]  # Add its weight
            uf.union(edges[e][0], edges[e][1])  # Connect its nodes

        # Try to build MST using Kruskal’s logic
        for i in range(len(edges)):

            # Skip blocked edge
            if i == block:
                continue

            u, v, w = edges[i][0], edges[i][1], edges[i][2]

            # If adding this edge forms a cycle → skip
            if uf.find_parent(u) == uf.find_parent(v):
                continue

            # Otherwise include edge in MST
            uf.union(u, v)
            weight += w

        # Check if all nodes are connected (valid MST)
        root = uf.find_parent(0)
        for i in range(n):
            if uf.find_parent(i) != root:
                return float('inf')  # Not connected → invalid MST

        return weight  # Return MST total weight